bring cache-dit to flux-fast

DefTruth opened the pull request:
## Install
docs: https://github.com/vipshop/cache-dit
```bash
pip install -U cache-dit
```

## BFloat16  

```bash
python run_benchmark.py \
    --ckpt "black-forest-labs/FLUX.1-dev" \
    --trace-file bfloat16.json.gz \
    --compile_export_mode disabled \
    --disable_fused_projections \
    --disable_channels_last \
    --disable_fa3 \
    --disable_quant \
    --disable_inductor_tuning_flags \
    --num_inference_steps 28 \
    --output-file output.png 
```

## BFloat16 + cache-dit

```bash
python run_benchmark.py \
    --ckpt "black-forest-labs/FLUX.1-dev" \
    --trace-file bfloat16.json.gz \
    --compile_export_mode disabled \
    --disable_fused_projections \
    --disable_channels_last \
    --disable_fa3 \
    --disable_quant \
    --disable_inductor_tuning_flags \
    --num_inference_steps 28 \
    --enable_cache_dit \
    --output-file output_cache.png 
``` 

## BFloat16 + cache-dit + torch.compile
```bash
# bf16 + cache-dit + torch.compile
python run_benchmark.py \
    --ckpt "black-forest-labs/FLUX.1-dev" \
    --trace-file bfloat16.json.gz \
    --compile_export_mode compile \
    --disable_fused_projections \
    --disable_channels_last \
    --disable_fa3 \
    --disable_quant \
    --disable_inductor_tuning_flags \
    --num_inference_steps 28 \
    --enable_cache_dit \
    --output-file output_cache_compile.png
```

## Metrics

|BF16|BF16 + cache-dit|BF16 + cache-dit + compile|
|:---:|:---:|:---:|
|Baseline|PSNR: 34.23|PSNR: 34.16|
|L20: 24.94s|L20: 20.85s|L20: 17.39s|
|![output](https://github.com/user-attachments/assets/4a9237c5-5736-483b-85f7-38ab6c417009)|![output_cache](https://github.com/user-attachments/assets/99b0abbc-3615-4e92-9b0f-c6c45ae6d24e)|![output_cache_compile](https://github.com/user-attachments/assets/f02243ed-4887-468d-874f-6e619af6d5cf)

```bash
cache-dit-metrics-cli all -i1 output.png -i2 output_cache.png
INFO 07-11 14:19:04 [metrics.py:415] output.png vs output_cache.png, Num: 1, PSNR: 34.232669211224206
INFO 07-11 14:19:05 [metrics.py:420] output.png vs output_cache.png, Num: 1, SSIM: 0.9748773813481453
INFO 07-11 14:19:05 [metrics.py:425] output.png vs output_cache.png, Num: 1,  MSE: 24.53654670715332
cache-dit-metrics-cli all -i1 output.png -i2 output_cache_compile.png
INFO 07-11 14:09:42 [metrics.py:415] output.png vs output_cache_compile.png, Num: 1, PSNR: 34.16003517257299
INFO 07-11 14:09:42 [metrics.py:420] output.png vs output_cache_compile.png, Num: 1, SSIM: 0.9749983003494543
INFO 07-11 14:09:43 [metrics.py:425] output.png vs output_cache_compile.png, Num: 1,  MSE: 24.950361569722492
```

DefTruth committed:
support cache-dit

sayakpaul reviewed (commented):
Nice, thanks for your contributions!

Would it be possible to also show some numbers on H100? Additionally, does `cache-dit` work with `torch.compile()`?

DefTruth committed:
torch.compile + cache-dit

sayakpaul commented:
Cool. Let's add some of these numbers to the README then?

DefTruth committed:
update README.md

sayakpaul reviewed (commented):
Left a couple more comments.

DefTruth committed:
disable cache-dit for experiments

DefTruth committed:
update README.md

DefTruth committed:
update README.md

DefTruth committed:
update README.md

DefTruth committed:
update README.md

DefTruth committed:
fix

DefTruth commented on README.md:
````diff
@@ -667,3 +668,31 @@ image = pipe(prompt, num_inference_steps=4).images[0]
+        )
+        pass
+```
+
````
> Sorry, I don't see those numbers. Where did you add them?

Added L20 results to README.md, please check `Cache Acceleration with cache-dit: DBCache + F12B12` part in https://github.com/xlite-dev/flux-fast/tree/cache-dit?tab=readme-ov-file#improvements-progressively section

<img width="1425" height="709" alt="image" src="https://github.com/user-attachments/assets/24c8b72a-33ae-4d37-bd67-429a0138fbfe" />

jbschlosser reviewed (commented):
Awesome, thanks for the contribution! I'll see if I can get some time to run some benchmarks with caching on my H100 machine

DefTruth committed:
fix

DefTruth commented on README.md:
```diff
@@ -44,6 +44,7 @@ Summary of the optimizations:
 * `torch.export` + Ahead-of-time Inductor (AOTI) + CUDAGraphs
+* Almost lossless cache acceleration with `cache-dit: DBCache + F12B12`
 
 All of the above optimizations are lossless (outside of minor numerical differences sometimes
```
@jbschlosser changed to:  

Cache Acceleration with `cache-dit: DBCache + F12B12`

DefTruth commented:
> Awesome, thanks for the contribution! I'll see if I can get some time to run some benchmarks with caching on my H100 machine

Feel free to update this PR and replace the result in L20 with the result from H100.

sayakpaul commented:
Will merge the PR after that. It would be good to investigate the graph break problems for potential improvements though. Do you want to open an issue in the cache-dit repo and tag us?

DefTruth commented:
> Will merge the PR after that. It would be good to investigate the graph break problems for potential improvements though. Do you want to open an issue in the cache-dit repo and tag us?

I have you open an issue in the cache-dit:

- https://github.com/vipshop/cache-dit/issues/76

DefTruth committed:
update .gitignore

DefTruth commented:
@sayakpaul @jbschlosser A relatively safe approach is to modify the `--disable_cache_dit (default: False)` option to `--enable_cache_dit (default: False)`. This way, under the default settings, all the original optimization options will not be affected. If you wish to do this, I can help with the modification.

DefTruth committed:
update README.md

DefTruth committed:
revert .gitignore

DefTruth committed:
change default flags

DefTruth commented on .gitignore:
```diff
@@ -1 +1,5 @@
+*.txt
+*.json.gz
+*.png
+black-forest-labs
```
reverted

DefTruth committed:
fix

DefTruth commented:
> @DefTruth I was also wondering if the cache-dit configuration options are hardware-dependent. If so, I think we should mention it in the README and provide guidance to the users about how they should tune it.
> 
> WDYT?

The configuration options for cache-dit are not hardware-dependent. I have already added a link to the documentation that provides guidance to users on how they should tune it.

DefTruth commented:
> Sure, let's do that. Thanks for willing to help.

Done~

DefTruth committed:
update README.md

DefTruth committed:
Update README.md

DefTruth committed:
Update README.md

sayakpaul reviewed (approved):
LGTM! Thanks for your hard work 🖖 

Will wait a bit for @jbschlosser to run H100 exps if possible before merging.

jbschlosser commented:
> Will wait a bit for @jbschlosser to run H100 exps if possible before merging.

just launched a set of experiments with caching enabled :)

jbschlosser commented:
ah one sec, looks like the flash_attn imports were messed up and ruined my runs

jbschlosser commented:
on H100 for Flux Schnell, I'm not seeing a huge difference with `--enable_cache_dit` and `cache_dit==0.2.8` installed:
```
results_mean_sec = [
    ("bfloat16", 1.1336731910705566),
    ("compile", 0.7163330912590027),
    ("qkv proj", 0.7130974531173706),
    ("channels_last", 0.6999137997627258),
    ("flash v3", 0.6080203652381897),
    ("float8", 0.4970979690551758),
    ("flags", 0.49017876386642456),
    ("export", "N/A"),
]
```

it doesn't work with `torch.export`, but I think that's expected due to the way `cache_dit` implements the Flux diffusers adapter. thoughts?

DefTruth commented:
> on H100 for Flux Schnell, I'm not seeing a huge difference with `--enable_cache_dit` and `cache_dit==0.2.8` installed:
> 
> ```
> results_mean_sec = [
>     ("bfloat16", 1.1336731910705566),
>     ("compile", 0.7163330912590027),
>     ("qkv proj", 0.7130974531173706),
>     ("channels_last", 0.6999137997627258),
>     ("flash v3", 0.6080203652381897),
>     ("float8", 0.4970979690551758),
>     ("flags", 0.49017876386642456),
>     ("export", "N/A"),
> ]
> ```
> 
> it doesn't work with `torch.export`, but I think that's expected due to the way `cache_dit` implements the Flux diffusers adapter. thoughts?

cache-dit is not suitable for Flux Schnell with only 4 steps (there's no need to use cache either)，please try Flux.1 Dev with 28 steps:   

```bash
# bf16
python run_benchmark.py \
    --ckpt "black-forest-labs/FLUX.1-dev" \
    --trace-file bfloat16.json.gz \
    --compile_export_mode disabled \
    --disable_fused_projections \
    --disable_channels_last \
    --disable_fa3 \
    --disable_quant \
    --disable_inductor_tuning_flags \
    --num_inference_steps 28 \
    --output-file output.png 

# bf16 + cache-dit + torch.compile
python run_benchmark.py \
    --ckpt "black-forest-labs/FLUX.1-dev" \
    --trace-file bfloat16.json.gz \
    --compile_export_mode compile \
    --disable_fused_projections \
    --disable_channels_last \
    --disable_fa3 \
    --disable_quant \
    --disable_inductor_tuning_flags \
    --num_inference_steps 28 \
    --enable_cache_dit \
    --output-file output_cache_compile.png
```
cache-dit doesn't work with `torch.export` now. cache-dit extends Flux and introduces some Python dynamic operations, so it may not be possible to export the model using torch.export.

DefTruth commented:
> flash_attn

weird, cache-dit is not depends on flash_attn.

Commit:
add warn to export

DefTruth commented:
> @jbschlosser
> 
> If you need, I can add a check to skip the export process when both cache-dit and export are enabled, and add a warning log to inform users that they need to disable cache-dit before using the export function.

@jbschlosser done~ PTAL

DefTruth commented:
@sayakpaul @jbschlosser more results on my main branch for L20, https://github.com/xlite-dev/flux-faster/tree/main?tab=readme-ov-file#flux-faster

# flux-faster

A forked version of [huggingface/flux-fast](https://github.com/huggingface/flux-fast) that **makes flux-fast even faster with [cache-dit](https://github.com/vipshop/cache-dit).**

```bash
pip3 install -U cache-dit # or: pip3 install git+https://github.com/vipshop/cache-dit.git
```
|BF16|BF16 + cache-dit|BF16 + cache-dit + compile|
|:---:|:---:|:---:|
|Baseline (FLUX.1-dev 28 steps)|PSNR: 34.23|PSNR: 34.16|
|L20: 24.94s|L20: 20.85s|L20: 17.39s|
|![output](https://github.com/user-attachments/assets/4a9237c5-5736-483b-85f7-38ab6c417009)|![output_cache](https://github.com/user-attachments/assets/99b0abbc-3615-4e92-9b0f-c6c45ae6d24e)|![output_cache_compile](https://github.com/user-attachments/assets/f02243ed-4887-468d-874f-6e619af6d5cf)|  
|BF16 + compile| BF16 + compile + qkv projection + channels_last + float8 quant + inductor flags |BF16 + compile + qkv projection + channels_last + float8 quant + inductor flags + **cache-dit**|
|PSNR: 19.28|PSNR: 18.07|PSNR: 22.24|
|L20: 20.24s|L20: 13.29s|L20: 11.21s|
|![bf16_compile](https://github.com/user-attachments/assets/a4bef05b-272d-42f7-ba02-6855731e6138)|![bf16_compile_qkv_chan_quant_flags](https://github.com/user-attachments/assets/5dc236e0-d2b4-43d1-bc60-ca9af51fd611)|![bf16_cache_compile_qkv_chan_quant_flags](https://github.com/user-attachments/assets/b96874a7-0224-41d7-be24-3f8810f9a4d3)|
|BF16 + **compile transformer block only** | BF16 + **compile transformer block only** + qkv projection + channels_last + float8 quant + inductor flags |BF16 + **compile transformer block only** + qkv projection + channels_last + float8 quant + inductor flags + **cache-dit**|
|PSNR: 39.72|PSNR: 21.77|PSNR: 21.89|
|L20: 20.49s|L20: 13.26s |L20: 11.14s|
|![bf16_compile_trn](https://github.com/user-attachments/assets/fa00a80f-b1cb-4c60-8bad-7fb369f0e280)|![bf16_compile_qkv_chan_quant_flags_trn](https://github.com/user-attachments/assets/958ae267-0351-4f85-b378-f863d9d3038c)|![bf16_cache_compile_qkv_chan_quant_flags_trn](https://github.com/user-attachments/assets/0ae9bf21-71d2-47d8-81e8-4fc7828dd801)|

## Important Notes

1) Please add `--enable_cache_dit` flag to use cache-dit. cache-dit doesn't work with torch.export now. cache-dit extends Flux and introduces some Python dynamic operations, so it may not be possible to export the model using torch.export.
2) Compiling the entire transformer appears to introduce precision loss in my tests on an NVIDIA L20 device (tested with PyTorch 2.7.1). Please try to add `--only_compile_transformer_blocks` flag to compile transformer blocks only if you want to keep higer precision.

## Experiments 

https://github.com/xlite-dev/flux-faster/blob/main/experiments_cache_ada.sh

DefTruth committed:
Update benchmark_utils.py

DefTruth committed:
Update pipeline_utils.py

DefTruth committed:
Update README.md

Commit:
export cache-dit configs

DefTruth committed:
Merge branch 'huggingface:main' into cache-dit

Commit:
fix typo

DefTruth committed:
Update README.md

DefTruth committed:
Update benchmark_utils.py

DefTruth committed:
Update README.md

DefTruth committed:
Update README.md

DefTruth committed:
Update benchmark_utils.py

DefTruth committed:
Update README.md

DefTruth committed:
Update pipeline_utils.py

DefTruth committed:
Update README.md

DefTruth committed:
Update README.md

DefTruth committed:
fix typo

jbschlosser commented on utils/pipeline_utils.py:
```diff
@@ -387,6 +414,34 @@ def optimize(pipeline, args):
+            print(
+                "Please install cache-dit via 'pip install -U cache-dit'"
+            )
+            pass
```
nit: I think it's best to error out entirely instead of swallowing this exception

DefTruth replied:
done

Commit:
better comments and messages

DefTruth committed:
better comments and messages

DefTruth commented:
> Good idea - I think it'd be nice to error out more smoothly if an unsupported configuration is encountered.

done~

DefTruth committed:
better comments and messages

DefTruth commented:
@jbschlosser hi~ can you take a look to the latest update?

sayakpaul commented on README.md:
```diff
@@ -139,31 +140,42 @@ Currently, only torch.export is not working as expected. Instead, use `torch.com
+                        [--residual_diff_threshold RESIDUAL_DIFF_THRESHOLD] 
+                        [--enable_taylorseer]
 
 options:
   -h, --help            show this help message and exit
-  --ckpt CKPT           Model checkpoint path (default: black-forest-labs/FLUX.1-schnell)
+  --ckpt {black-forest-labs/FLUX.1-schnell,black-forest-labs/FLUX.1-dev,black-forest-labs/FLUX.1-Kontext-dev}
+                        Model checkpoint path (default: black-forest-labs/FLUX.1-schnell)
   --prompt PROMPT       Text prompt (default: A cat playing with a ball of yarn)
+  --image IMAGE         Image to use for Kontext (default: None)
   --cache-dir CACHE_DIR
-                        Cache directory for storing exported models (default:
-                        ~/.cache/flux-fast)
+                        Cache directory for storing exported models (default: /root/.cache/flux-fast)
+  --use-cached-model    Attempt to use cached model only (don't re-export) (default: False)
```
Why these changes?

sayakpaul commented on utils/pipeline_utils.py:
```diff
@@ -412,14 +479,20 @@ def optimize(pipeline, args):
-            serialize=(not args.use_cached_model),
-            is_timestep_distilled=is_timestep_distilled
-        )
+        if not args.enable_cache_dit:
```
I think we can club all the checks with regards to the parsed arguments in a common utility called `validate_args()`. This way, the code will be cleaner and easier to read. WDYT?

sayakpaul reviewed (commented):
Thanks for your continuous efforts. I left another round of feedback.

sayakpaul commented on utils/pipeline_utils.py:
```diff
@@ -387,6 +417,43 @@ def optimize(pipeline, args):
+    # cache-dit: DBCache configs
+    if args.enable_cache_dit:
+        if args.ckpt.endswith("schnell"):
+            print(
+                "cache-dit is not suitable for FLUX.1-schnell with only 4 steps "
+                "(there's no need to use cache either), please try FLUX.1-dev "
+                "or FLUX.1-Kontext-dev with 28 steps."
+            )
```
```suggestion
            raise ValueError(
                "cache-dit is not suitable for FLUX.1-schnell with only 4 steps "
                "(there's no need to use cache either), please try FLUX.1-dev "
                "or FLUX.1-Kontext-dev with 28 steps."
            )
```

This is what I am suggesting.

DefTruth replied:
copy that~

DefTruth commented on README.md:
```diff
@@ -139,31 +140,42 @@ Currently, only torch.export is not working as expected. Instead, use `torch.com
+                        [--compile_export_mode {compile,export_aoti,disabled}] 
+                        [--only_compile_transformer_blocks] [--disable_fused_projections] 
+                        [--disable_channels_last] [--disable_fa3] [--disable_quant]
+                        [--disable_inductor_tuning_flags] [--enable_cache_dit] 
+                        [--Fn_compute_blocks FN_COMPUTE_BLOCKS] 
+                        [--Bn_compute_blocks BN_COMPUTE_BLOCKS] 
+                        [--warmup_steps WARMUP_STEPS]
+                        [--max_cached_steps MAX_CACHED_STEPS] 
+                        [--residual_diff_threshold RESIDUAL_DIFF_THRESHOLD] 
+                        [--enable_taylorseer]
```
good idea~

DefTruth replied:
done~, like this:

```python
from cache_dit.cache_factory import apply_cache_on_pipe
from cache_dit.cache_factory import load_cache_options_from_yaml
cache_options = load_cache_options_from_yaml(
       args.cache_dit_config
)
apply_cache_on_pipe(pipeline, **cache_options)
```

DefTruth committed:
update README.md

DefTruth commented on utils/pipeline_utils.py:
```diff
@@ -231,14 +231,44 @@ def wrapped(*args, **kwargs):
+        torch._dynamo.config.accumulated_recompile_limit = (
+            2048  # default is 256
+        )
+        for module in pipeline.transformer.transformer_blocks:
+            module.compile(
+                mode="max-autotune-no-cudagraphs", 
+                dynamic=True if is_hip() else None
+            )
+        for module in pipeline.transformer.single_transformer_blocks:
+            module.compile(
+                mode="max-autotune-no-cudagraphs", 
+                dynamic=True if is_hip() else None
+            )
```
it seems that i can try:  

```python
pipe.transformer.compile(use_regional_compile=True/False, **kwargs)
```

DefTruth replied:
Done~ Temporarily remove this option until I have more comprehensive tests before considering submitting a new pr for modification.

DefTruth committed:
update pipeline_utils.py

jbschlosser reviewed (approved):
Thanks for all the hard work! looks great to me once the outstanding feedback is addressed :)

DefTruth commented:
I will also update the code today to fix the issues mentioned in this round of reviews. Please wait until I have completed the improvements before considering merging it.

DefTruth committed:
load cache_config from yaml

DefTruth commented on README.md:
```diff
@@ -139,31 +140,42 @@ Currently, only torch.export is not working as expected. Instead, use `torch.com
-                        export_aoti)
+                        Configures how torch.compile or torch.export + AOTI are used (default: export_aoti)
+  --only_compile_transformer_blocks
+                        Only compile Transformer Blocks for higher precision (default: False)
```
Done~ Temporarily remove this option until I have more comprehensive tests before considering submitting a new pr for modification.

DefTruth commented:
@sayakpaul @jbschlosser PTAL to the **New commits**:

- Simplified cache-dit flags and load cache configurations from yaml
- Temporarily removed the `--only_compile_transformer_blocks` option; I will consider submitting a new PR for modification after conducting more comprehensive tests.
- Simplified the [experiments_cache.sh](https://github.com/xlite-dev/flux-faster/blob/main/experiments_cache.sh) script. If you want to include this script in flux-fast, I can add it (but I don't have an H100 device to test fa3 + cache-dit).

## Experiments  

Please run [experiments_cache.sh](https://github.com/xlite-dev/flux-faster/blob/main/experiments_cache.sh) script to reproduce the results. For example:

```bash
#!/bin/bash

# TODO: FA3, H100
CKPT="black-forest-labs/FLUX.1-dev"

# bfloat16
python run_benchmark.py \
    --ckpt ${CKPT} \
    --trace-file bf16.json.gz \
    --compile_export_mode disabled \
    --disable_fused_projections \
    --disable_channels_last \
    --disable_fa3 \
    --disable_quant \
    --disable_inductor_tuning_flags \
    --num_inference_steps 28 \
    --output-file bf16.png \
    > bf16.txt 2>&1

# bfloat16 + torch.compile + qkv projection + channels_last + float8 quant + inductor flags
python run_benchmark.py \
    --ckpt ${CKPT} \
    --trace-file optimized.json.gz \
    --compile_export_mode compile \
    --disable_fa3 \
    --num_inference_steps 28 \
    --output-file optimized.png \
    > optimized.txt 2>&1

# bfloat16 + torch.compile + qkv projection + channels_last + float8 quant + inductor flags 
# + cache_dit
python run_benchmark.py \
    --ckpt ${CKPT} \
    --trace-file optimized_cache_dit.json.gz \
    --compile_export_mode compile \
    --disable_fa3 \
    --num_inference_steps 28 \
    --enable_cache_dit \
    --cache_dit_config cache_config.yaml \
    --output-file optimized_cache_dit.png \
    > optimized_cache_dit.txt 2>&1
```

## PSNR

cache-dit is very fast and accurate! (tested on NVIDIA L20)

```bash
cache-dit-metrics-cli psnr -i1 bf16.png -i2 optimized.png
INFO 07-17 13:37:52 [metrics.py:475] bf16.png vs optimized.png, Num: 1, PSNR: 18.12911771154573
cache-dit-metrics-cli psnr -i1 bf16.png -i2 optimized_cache_dit.png
INFO 07-17 13:38:01 [metrics.py:475] bf16.png vs optimized_cache_dit.png, Num: 1, PSNR: 24.537262573744506
```

sayakpaul commented on README.md:
```diff
@@ -139,31 +140,33 @@ Currently, only torch.export is not working as expected. Instead, use `torch.com
+                        [--output-file OUTPUT_FILE] [--seed SEED] [--trace-file TRACE_FILE] [--disable_bf16]
+                        [--compile_export_mode {compile,export_aoti,disabled}] [--disable_fused_projections] 
+                        [--disable_channels_last] [--disable_fa3] [--disable_quant] [--disable_inductor_tuning_flags] 
+                        [--enable_cache_dit] [--cache_dit_config CACHE_DIT_CONFIG]
```
I think if `cache_dit_config` is provided (which is defaulted to None) should be enough for us to determine if `cache_dit` is enabled. So, no need to have a different flag as `enable_cache_dit`. WDYT?

DefTruth replied:
done

sayakpaul commented on README.md:
```diff
@@ -172,6 +175,9 @@ options:
   --disable_quant       Disables usage of dynamic float8 quantization (default: False)
   --disable_inductor_tuning_flags
                         Disables use of inductor tuning flags (default: False)
+  --enable_cache_dit    Enables use of cache-dit: DBCache (default: False)
+  --cache_dit_config CACHE_DIT_CONFIG
+                        Cache options config of cache-dit: DBCache (default: cache_config.yaml)
```
Same as above.

sayakpaul commented on utils/benchmark_utils.py:
```diff
@@ -54,6 +54,11 @@ def create_parser():
     parser.add_argument("--disable_inductor_tuning_flags", action="store_true",
                         help="Disables use of inductor tuning flags")
+    # cache-dit: DBCache configs
+    parser.add_argument("--enable_cache_dit", action="store_true",
+                        help="Enables use of cache-dit: DBCache")
+    parser.add_argument("--cache_dit_config", type=str, default="cache_config.yaml",
+                        help="Cache options config of cache-dit: DBCache")
```
Same as above.

sayakpaul commented on utils/pipeline_utils.py:
```diff
@@ -387,6 +396,31 @@ def optimize(pipeline, args):
         pipeline.vae.to(memory_format=torch.channels_last)
+    
+    # cache-dit: DBCache configs
+    if args.enable_cache_dit:
```
```suggestion
    if args.cache_dit_config is not None:
```

sayakpaul commented on utils/pipeline_utils.py:
```diff
@@ -387,6 +396,31 @@ def optimize(pipeline, args):
+            # docs: https://github.com/vipshop/cache-dit
+            from cache_dit.cache_factory import apply_cache_on_pipe
+            from cache_dit.cache_factory import load_cache_options_from_yaml
+            cache_options = load_cache_options_from_yaml(
+                args.cache_dit_config
+            )
```
It's a simple yaml load right? If so, prefer more explicit code:

```py
import yaml

with open(file_path, "r") as f: config = yaml.safe_load(f)
```

DefTruth replied:
not a simple yaml load, please check: 

https://github.com/vipshop/cache-dit/blob/ff6b37f6b9e0598d8a6329379d55042f4190ddac/src/cache_dit/cache_factory/utils.py#L5

sayakpaul reviewed (approved):
Thanks. Just another set of reviews.

DefTruth committed:
fix

DefTruth committed:
fix

DefTruth commented:
@sayakpaul  fixed

sayakpaul reviewed (approved):
Let's go!

sayakpaul commented:
Thanks a lot for your patience and for bearing with the reviews. Let's try fixing the graph break issues and see how far can we push this!
